src/test/java/org/apache/sling/models/testmodels/classes/CachedModelWithSelfReference.java

```diff
@@ -22,304 +22,14 @@
-        l3637, l3638, l3639, l3640, l3641, l3642, l3643, l3644, l3645, l3646, l3647, l3648, l3649,
-        l3650, l3651, l3652, l3653, l3654, l3655, l3656, l3657, l3658, l3659, l3660, l3661, l3662,
-        l3663, l3664, l3665, l3666, l3667, l3668, l3669, l3670, l3671, l3672, l3673, l3674, l3675,
-        l3676, l3677, l3678, l3679, l3680, l3681, l3682, l3683, l3684, l3685, l3686, l3687, l3688,
-        l3689, l3690, l3691, l3692, l3693, l3694, l3695, l3696, l3697, l3698, l3699, l3700, l3701,
-        l3702, l3703, l3704, l3705, l3706, l3707, l3708, l3709, l3710, l3711, l3712, l3713, l3714,
-        l3715, l3716, l3717, l3718, l3719, l3720, l3721, l3722, l3723, l3724, l3725, l3726, l3727,
-        l3728, l3729, l3730, l3731, l3732, l3733, l3734, l3735, l3736, l3737, l3738, l3739, l3740,
-        l3741, l3742, l3743, l3744, l3745, l3746, l3747, l3748, l3749, l3750, l3751, l3752, l3753,
-        l3754, l3755, l3756, l3757, l3758, l3759, l3760, l3761, l3762, l3763, l3764, l3765, l3766,
-        l3767, l3768, l3769, l3770, l3771, l3772, l3773, l3774, l3775, l3776, l3777, l3778, l3779,
-        l3780, l3781, l3782, l3783, l3784, l3785, l3786, l3787, l3788, l3789, l3790, l3791, l3792,
-        l3793, l3794, l3795, l3796, l3797, l3798, l3799, l3800, l3801, l3802, l3803, l3804, l3805,
-        l3806, l3807, l3808, l3809, l3810, l3811, l3812, l3813, l3814, l3815, l3816, l3817, l3818,
-        l3819, l3820, l3821, l3822, l3823, l3824, l3825, l3826, l3827, l3828, l3829, l3830, l3831,
-        l3832, l3833, l3834, l3835, l3836, l3837, l3838, l3839, l3840, l3841, l3842, l3843, l3844,
-        l3845, l3846, l3847, l3848, l3849, l3850, l3851, l3852, l3853, l3854, l3855, l3856, l3857,
-        l3858, l3859, l3860, l3861, l3862, l3863, l3864, l3865, l3866, l3867, l3868, l3869, l3870,
-        l3871, l3872, l3873, l3874, l3875, l3876, l3877, l3878, l3879, l3880, l3881, l3882, l3883,
-        l3884, l3885, l3886, l3887, l3888, l3889, l3890, l3891, l3892, l3893, l3894, l3895, l3896,
-        l3897, l3898, l3899, l3900, l3901, l3902, l3903, l3904, l3905, l3906, l3907, l3908, l3909,
-        l3910, l3911, l3912, l3913, l3914, l3915, l3916, l3917, l3918, l3919, l3920, l3921, l3922,
-        l3923, l3924, l3925, l3926, l3927, l3928, l3929, l3930, l3931, l3932, l3933, l3934, l3935,
-        l3936, l3937, l3938, l3939, l3940, l3941, l3942, l3943, l3944, l3945, l3946, l3947, l3948,
-        l3949, l3950, l3951, l3952, l3953, l3954, l3955, l3956, l3957, l3958, l3959, l3960, l3961,
-        l3962, l3963, l3964, l3965, l3966, l3967, l3968, l3969, l3970, l3971, l3972, l3973, l3974,
-        l3975, l3976, l3977, l3978, l3979, l3980, l3981, l3982, l3983, l3984, l3985, l3986, l3987,
-        l3988, l3989, l3990, l3991, l3992, l3993, l3994, l3995, l3996, l3997, l3998, l3999, l4000;
+    // Add a big array of longs to the class, to make it take fewer iterations to fill the heap
+    @SuppressWarnings("unused")
+    long[] longs = new long[numberOfLongs];
```

Comment:
are you sure this isn't optimized away by the compiler?

Can you actually trigger an OOM if you don't cache this model?

Reply:
I checked on the commit before this fix/test was introduced, and it did OOM (at least on JDK 8). When testing after the fix, it passed (as expected).

 If we want to be absolutely sure, I can add something into the test that accesses the array, and make the array `public`. That may guard us until the JDK gets some super-optimization powers in the future. That being said, usually these tests don't run in a "warmed" JVM (as far as I can tell), so I doubt we even get to C1, let alone C2 optimizations. Most likely, this code is executed interpreted.

Reply:
Thinking some more on it, I sincerely doubt anything around the array is being optimized out. This is all thanks to the caching that this test is attempting to exercise in the first place. In the test, it is just looping `maxInstances` times and discarding the return value:

```java
        for (long i = 0; i < maxInstances; i++) {
            factory.createModel(mock(SlingHttpServletRequest.class), CachedModelWithSelfReference.class);
        }
```

If something were to be optimized out, it would likely be the call to `createModel`, but because of the "impurity" of `createModel` implementation (aka it changes state in the model factory), the JVM cannot optimize it out. The JVM, at least based on OpenJDK, has no means of eliminating object state as a micro-optimization that I am aware of.

That being said, I am all for protecting ourselves in the manner mentioned in my previous comment. The down side would be a slight slowing of the test due to the extra logic (though that would not be the dominating factor in the performance).

Disclaimer: I am far from a JVM expert, but I do follow several OpenJDK mailing lists and have learned quite a bit about how C1/C2 optimize stuff by following those lists & reading [Aleksey Shipilëv's blog][1]

[1]: https://shipilev.net/